packages/endpoint/package.json

```
@@ -25,23 +25,40 @@
       ]
     }
   },
+  "exports": {
+    ".": [
+      {
+        "node": {
+          "import": "./node.mjs",
+          "require": "./dist/index.cjs"
+        },
```

the "node" nesting isn't really needed; you could just have "import" and "default" and it'd be fine.
```suggestion
        "import": "./node.mjs",
```

---

because node supports both cjs and modules, it is important to route node to the same way no matter how the import happens as node can potentially import the same package in both ways. there are several solutions to this, but this one does it by routing all node loads to cjs.

---

I'm not really sure what you mean. with "import" and then "default", node will always import the "import" file (there's only one way to import anything). The "node" nesting only matters if there existed some tool that respected something _else_ - which does not exist.

---

From the previous PR:

References for behavior in node, webpack and typescript (not supported yet)
- https://webpack.js.org/guides/package-exports
- https://nodejs.org/docs/latest-v15.x/api/packages.html
- https://github.com/microsoft/TypeScript/issues/33079

---

I have no idea why these docs authored this, but as a member of the node modules group that defined how it works, I'm telling you this is unnecessary complexity that achieves nothing.

---

I've been able to reproduce the [dual package hazard](https://nodejs.org/docs/latest-v15.x/api/packages.html#packages_approach_1_use_an_es_module_wrapper) on my own. This is more about how commonjs + es6 modules work together than exports.

---

This configuration will have no impact on that hazard - it’s exactly the same, but with more steps.

---

I don't follow where you are deriving this conclusion from. Using both es6 module import and commonjs require will route to the same file, which means members like classes will be ===.

---

With the PR (with the "node" object):
A require of the package will hit the ".", then the "node", then the "require", going to `dist/index.cjs`.
An import of the package will hit the ".", then the "node", then the "import", going to `./node.mjs`.

With my suggestion (without the "node" object):
A require of the package will hit the ".", then the "default", going to `dist/index.cjs`.
An import of the package will hit the ".", then the "import", going to `./node.mjs`.

So, regardless of what `node.mjs` does, both approaches are *identical*, but yours has extra steps.

The way to avoid the dual package hazard is to ensure that `node.mjs` imports any state/singletons from the same CJS file that `dist/index.cjs` does - this is orthogonal to either scenario above.

---

With the PR (with the "node" object):
A require of the package will hit the ".", then the "node", then the "require", going to `dist/index.cjs`.
An import of the package will hit the ".", then the "node", then the "import", going to `./node.mjs`.

With my suggestion (without the "node" object):
A require of the package will hit the ".", then the "default", going to `dist/index.cjs`.
An import of the package will hit the ".", then the "import", going to `./node.mjs`.

So, regardless of what `node.mjs` does, both approaches are *identical*, but yours has extra steps.

The way to avoid the dual package hazard is to ensure that `node.mjs` imports any state/singletons from the same CJS file that `dist/index.cjs` does - this is orthogonal to either scenario above.

---

Removing node as you say will still work for node, yes. However, it will serve the cjs bundle to tools like webpack, which is not what we want. CJS is built with babel settings targetting node, vs the lib are targetting browsers. Furthermore, CJS does not support tree shaking, etc, which does work with webpack. (And yes, validated in practice)

---

So you're saying that, despite `. -> import` pointing to the ESM file, webpack needs `. -> node -> import` to be able to find it?

(ftr, CJS absolutely supports treeshaking, precisely as well as ESM; webpack and rollup are just too lazy to implement it)

---

The design and validation is covered in https://github.com/coinbase/rest-hooks/pull/590. This is simply applying the previously validated design to another package.

Here's a summary:

```
  "exports": {
    // this is the export without any path
    ".": [
      {
        // node has its own files for two reasons - they are in commonjs and transpiled to target node 12
        "node": {
          // node world understands these files so this helps maximize compat with this
          // because a given node can pull both the es6 module and commonjs version, this file simply redirects to the cjs file
          "import": "./node.mjs",
           // this is the commonjs version directly
          "require": "./dist/index.cjs"
        },
       // for maximum future or unknown compat, we provide a default case that uses ESM. this targets jan 2020 browsers
        "default": "./lib/index.js"
      },
      // this is fallback in case a tool doesn't understand conditionals
      "./lib/index.js"
    ],
   // tools like importing package.json, but all / imports are explicitly needed
    "./package.json": "./package.json"
  },
```

---

Complexity is set at 4:

- unpkg: I hear this is a thing. Don't use it myself, but I can see its use
- node
- modern module tooling like webpack
- legacy module tooling like webpack 4

---

OK - i now see the `./lib/index.js` default, versus the `./dist/index.cjs` node.require - i'd missed that.

Anything that doesn't understand exports conditions would also not understand ESM, so I'm still confused why this is here. Webpack understands those conditions - so wouldn't it select the ESM version regardless?

---

I'm not sure of specific tools that have this limitation, but it literally just adds another line to the package.json.  We have the same complexity in how many outputs we build. Is it really that problematic to be safe here? In the future this can be removed if it's proved absolutely unnnecessary